src/box/lua/trigger.c

```diff
@@ -253,3 +253,146 @@ box_lua_trigger_init(struct lua_State *L)
+		    luaL_iscallable(L, bottom));
+	if (!ok)
+		luaL_error(L, "trigger reset: incorrect arguments");
+}
+
+/**
+ * Sets or deletes trigger by name depending on passed arguments. Value at
+ * name_idx must be a string, value at func_idx must be a callable object,
+ * nil or box.NULL. Otherwise, an error will be thrown.
+ */
+static int
+luaT_event_reset_trigger_by_name(struct lua_State *L, struct event *event,
+				 int name_idx, int func_idx)
+{
+	if (lua_type(L, name_idx) != LUA_TSTRING)
+		luaL_error(L, "name must be a string");
+	const char *trigger_name = lua_tostring(L, name_idx);
+	if (luaL_iscallable(L, func_idx)) {
+		struct func_adapter *func =
+			func_adapter_lua_create(L, func_idx);
+		event_reset_trigger(event, trigger_name, func);
+		lua_pushvalue(L, func_idx);
+		return 1;
+	} else if (lua_isnil(L, func_idx) || luaL_isnull(L, func_idx)) {
+		event_reset_trigger(event, trigger_name, NULL);
+		return 0;
+	}
+	return luaL_error(L, "func must be a callable object or nil");
+}
+
+int
+luaT_event_reset_trigger(struct lua_State *L, int bottom, struct event *event)
+{
+	assert(L != NULL);
+	assert(bottom >= 1);
+	assert(event != NULL);
+	/* Use key-value API if the first argument is a non-callable table. */
+	if (lua_gettop(L) == bottom && lua_istable(L, -1) &&
+	    !luaL_iscallable(L, -1)) {
+		lua_getfield(L, bottom, "name");
+		lua_getfield(L, bottom, "func");
+		return luaT_event_reset_trigger_by_name(L, event, -2, -1);
+	}
+	/* Old way with name support. */
+	luaT_event_reset_trigger_check_positional_input(L, bottom);
+	const int top = bottom + 2;
+	if (!lua_isnil(L, top) && !luaL_isnull(L, top))
+		return luaT_event_reset_trigger_by_name(L, event, top, bottom);
+	/*
+	 * Name is not passed - old API support.
+	 * 1. If triggers are not passed, return table of triggers.
+	 * 2. If new_trigger is passed and old_trigger is not - set
+	 * new_trigger using string representation of function as name.
+	 * 3. If old_trigger is passed and new_trigger is not - delete
+	 * trigger with by function string representation as a name.
+	 * 4. If both triggers are provided - replace old trigger with
+	 * new one if they have the same name, delete old trigger and
+	 * insert new one at the beginning of the trigger list otherwise.
+	 */
+	if (!luaL_iscallable(L, bottom) && !luaL_iscallable(L, bottom + 1)) {
+		lua_createtable(L, 0, 0);
+		const char *name = NULL;
+		struct func_adapter *trigger = NULL;
+		struct event_trigger_iterator it;
+		event_trigger_iterator_create(&it, event);
+		int idx = 0;
+		while (event_trigger_iterator_next(&it, &trigger, &name)) {
+			idx++;
+			func_adapter_lua_get_func(trigger, L);
+			lua_rawseti(L, -2, idx);
+		}
+		event_trigger_iterator_destroy(&it);
+		return 1;
+	}
+	int ret_count = 0;
+	const char *old_name = NULL;
+	const char *new_name = NULL;
+	struct func_adapter *new_trg = NULL;
+	if (luaL_iscallable(L, bottom + 1)) {
+		old_name = luaT_tolstring(L, bottom + 1, NULL);
```

Comment:
It's not entirely correct to use `tostring` on a callable object because it may have the `__tostring` metamethod overwritten, in which case it'd return the same value even for different objects. You should use `lua_topointer` here.

Reply:
Good point, thanks. Use handler address now.